### ISSUE TITLE
Process node/ponder gets stuck in runtime

### ISSUE THREAD

**Issue by samclassix** [Frankencoin-ZCHF/ponder#2]
## Process node/ponder gets stuck in runtime

**Affected: All branches (?)**

#### What happened? What could be the issue?

Ponder indexer gets stuck and not responding anymore. it's possible that the "Equity:Delegation" code section is causing an infinite loop or a recursive update cycle that leads to resource exhaustion and causes the Ponder process to get stuck.

#### Posible code section

```
	const addParentDelegation = async (delegatedFrom: string | undefined, arrayToAdd: string[]) => {
		if (!delegatedFrom) return;

		const parentDelegation = await Delegation.upsert({
			id: delegatedFrom,
			create: {
				owner: delegatedFrom,
				pureDelegatedFrom: arrayToAdd,
			},
			update: ({ current }) => ({
				pureDelegatedFrom: current.pureDelegatedFrom.concat(
					arrayToAdd.filter((item) => current.pureDelegatedFrom.indexOf(item) < 0)
				),
			}),
		});

		addParentDelegation(parentDelegation.delegatedTo, arrayToAdd); // <<<<< <<<<< <<<<<
	};
```

#### Excluded code section, to don't crash ponder

```
	// await Delegation.update({
	//   id: event.args.from,
	//   data: {
	//     delegatedTo: event.args.to,
	//   },
	// });

	// addParentDelegation(event.args.to, arrayToUpdate);

	// await ActiveUser.upsert({
	//   id: event.args.from,
	//   create: {
	//     lastActiveTime: event.block.timestamp,
	//   },
	//   update: () => ({
	//     lastActiveTime: event.block.timestamp,
	//   }),
	// });
```

#### AI: Cody Analysis 

Here's a potential scenario that could cause such an issue:

When the Equity:Delegation event is triggered, the code updates the Delegation table for the event.args.from address.
The commented-out code attempts to update the delegatedTo field in the Delegation table for event.args.from with event.args.to.
The addParentDelegation function is then called with event.args.to and arrayToUpdate (which contains event.args.from).
Inside addParentDelegation, if event.args.to has a delegation record, it might update the pureDelegatedFrom array for that record, adding event.args.from to it.
If the update to the pureDelegatedFrom array triggers another Equity:Delegation event (which is possible if there's a listener for changes to the Delegation table), it could cause the same code to run again, leading to a recursive loop.
This recursive loop could potentially continue indefinitely, with each iteration updating the Delegation table and triggering more events, eventually exhausting system resources like memory or CPU.

To prevent such issues, it's essential to carefully handle recursive updates and ensure that there are no infinite loops or cycles in the event handling logic. Some potential solutions could be:

Implement proper checks and guards to prevent recursive updates or infinite loops.
Use a queue or a job system to process events asynchronously, preventing recursive calls within the same event handler.
Implement rate-limiting or throttling mechanisms to prevent excessive updates or events from overwhelming the system.
Thoroughly test the event handling logic with various edge cases and scenarios to identify and fix any potential issues before deploying to production.
It's also worth noting that Ponder is designed to handle events efficiently, so if the issue persists even after implementing safeguards, it might be worth investigating if there are any underlying issues or inefficiencies in the event handling logic or the database operations.

**Comment by samclassix** [comment]
### Dave from event (triggered)

**args: from event**
**delegation: from context.db**


```
{
  args: {
    from: '0xd872b7d4889Ab1c7de5a17a7877cA641AD2c3F72',
    to: '0x0C47e6C912Fe14a547055E8647cA2f97FF4eFA20'
  },
  delegation: {
    id: '0xd872b7d4889Ab1c7de5a17a7877cA641AD2c3F72',
    owner: '0xd872b7d4889Ab1c7de5a17a7877cA641AD2c3F72',
    pureDelegatedFrom: [],
    delegatedTo: null
  },
  arrayToUpdate: [ '0xd872b7d4889Ab1c7de5a17a7877cA641AD2c3F72' ]
}
{
  args: {
    from: '0x14c1C28cBD5C713DB0f58D51d2550c871D07C641',
    to: '0x14c1C28cBD5C713DB0f58D51d2550c871D07C641'
  },
  delegation: {
    id: '0x14c1C28cBD5C713DB0f58D51d2550c871D07C641',
    owner: '0x14c1C28cBD5C713DB0f58D51d2550c871D07C641',
    pureDelegatedFrom: [],
    delegatedTo: null
  },
  arrayToUpdate: [ '0x14c1C28cBD5C713DB0f58D51d2550c871D07C641' ]
}
{
  args: {
    from: '0x024580A265422Dd58aCB2a6F77300C74c095ea91',
    to: '0x024580A265422Dd58aCB2a6F77300C74c095ea91'
  },
  delegation: {
    id: '0x024580A265422Dd58aCB2a6F77300C74c095ea91',
    owner: '0x024580A265422Dd58aCB2a6F77300C74c095ea91',
    pureDelegatedFrom: [],
    delegatedTo: null
  },
  arrayToUpdate: [ '0x024580A265422Dd58aCB2a6F77300C74c095ea91' ]
}
{
  args: {
    from: '0xf94873c609B5fbEe3552cb23B80E53Ae209d5E49',
    to: '0xf94873c609B5fbEe3552cb23B80E53Ae209d5E49'
  },
  delegation: {
    id: '0xf94873c609B5fbEe3552cb23B80E53Ae209d5E49',
    owner: '0xf94873c609B5fbEe3552cb23B80E53Ae209d5E49',
    pureDelegatedFrom: [],
    delegatedTo: null
  },
  arrayToUpdate: [ '0xf94873c609B5fbEe3552cb23B80E53Ae209d5E49' ]
}
```

### 4 events

#### from
0xd872b7d4889Ab1c7de5a17a7877cA641AD2c3F72
**0x14c1C28cBD5C713DB0f58D51d2550c871D07C641
0x024580A265422Dd58aCB2a6F77300C74c095ea91
0xf94873c609B5fbEe3552cb23B80E53Ae209d5E49**

# The last three events delegate to themselves, does this create a infinite loop or a recursive update cycle?

**Comment by samclassix** [comment]
https://github.com/Frankencoin-ZCHF/ponder/tree/dev-roll-back

### Simple fix
simply "upsert" or "update" delegation in table by updating the "delegatedTo" value.
Schema:
```
Delegation: p.createTable({
        id: p.string(),
        owner: p.string(),
        delegatedTo: p.string(),
}),
```


### If you need a list of all delegations to a specific address 
use this ponder query in your client or app:

```
query {
		delegations (where: {delegatedTo: "0x0C47e6C912Fe14a547055E8647cA2f97FF4eFA20"}) {
			items {
				owner
                                  delegatedTo
			}
		}
	}
```

**Comment by samclassix** [comment]
I am not sure how you get that "main" finishes indexing without crashing. I index from zero, and as soon the indexer hits:
```
8:37:25 AM INFO  indexing   Indexed 1 Equity:Delegation event (chainId=1 block=19620385 logIndex=426)
```
it gets stuck and not responsive anymore, never successfully finishing up the initial indexing.

![Screenshot 2024-05-22 at 10 38 29 AM](https://github.com/Frankencoin-ZCHF/ponder/assets/96245183/de962a1e-e2e9-4ae6-b0f5-3e9231d6001b)

**Comment by luziusmeisser** [comment]
I don't know. When restarting, it does not re-index everything. Maybe it worked with an earlier version and then after a change it did not re-index all the events again? I would say: we fix the bug, deploy the fix, and then we check whether the data returned from mainnetponder.frankencoin.com is actually correct with regards to this event.

**Comment by samclassix** [comment]
Merged Dev roll back #3 https://github.com/Frankencoin-ZCHF/ponder/pull/3 into dev, which fixes the indexing by simplifying the delegation into:

```
Delegation: p.createTable({
        id: p.string(),
        owner: p.string(),
        delegatedTo: p.string(),
}),
```

i am not sure if this will break anything on the dapp, however, ponders branch dev works so far with dapp branch dev. 
Maybe there could be some issues when it come to governance. However, there is not much implemented on the dapp, that we could take the risk to test this on the deployed version on https:devapp.frankencoin.com
